enhancement(host_metrics source): Add initial cgroups metrics

bruceg opened the pull request:
This is just an initial implementation. Many more statistics are available, but these are the first ones that were requested. The framework should make adding more relatively straightforward.

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

Closes #5037

bruceg committed:
enhancement(host_metrics source): Add initial cgroups metrics

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

jszwedko commented on website/cue/reference/components/sources/host_metrics.cue:
```diff
@@ -116,6 +117,72 @@ components: sources: host_metrics: {
 		}
+		cgroups: {
+			common:      false
+			description: #"Options for the "cgroups" metrics collector. This collector is only available on Linux systems, as the cgroups (control groups) system is only present on Linux."#
```
I think we should make it clear that we only support cgroups v2 / unified (at least for now).

bruceg replied:
Agreed. Addressed in 2e4c2a271

tobz commented on src/sources/host_metrics/cgroups.rs:
```diff
@@ -0,0 +1,271 @@
+            ));
+        }
+
+        if !cgroup.is_root() {
+            if let Some(current) = filter_result_sync(
+                cgroup.load_memory_current(),
```
Take it or leave it: this almost feels like it should be encapsulated in `CGroup::load_memory_current` itself, by returning `io::Result<Option<MemoryStat>>` to account vs root vs non-root groups.

bruceg replied:
That's worth considering, and I was on the fence myself. All of these parsing routines should likely be handling `ENOENT` as distinct from other error cases, so that would be a model to follow for all related functions. On the other hand, those memory files are explicitly present in all non-root groups, so `ENOENT` is potentially an error worth reporting instead of just silently handling as no metrics present.

tobz reviewed (approved):
Tentatively approved.

Overall, it feels maybe a touch too under-documented: even if we simply linked to the `cgroups` documentation in the kernel, that would be helpful to understand how it's working and why it's doing things the way it's doing them.

tobz commented on src/sources/host_metrics/cgroups.rs:
```diff
@@ -0,0 +1,271 @@
+}
+
+impl CGroup {
+    fn root<P: AsRef<Path>>(base_group: Option<P>) -> Option<CGroup> {
+        let base_dir = join_path(heim::os::linux::sysfs_root(), "fs/cgroup");
+        let root = join_path(&base_dir, "unified");
+        is_dir(&root)
+            .then(|| root)
+            .or_else(|| is_file(join_path(&base_dir, "cgroup.procs")).then(|| base_dir))
+            .and_then(|root| match base_group {
+                Some(group) => {
+                    let group = group.as_ref();
+                    let root = join_path(root, group);
+                    is_dir(&root).then(|| CGroup {
+                        root,
+                        name: group.into(),
+                    })
+                }
+                None => Some(CGroup {
+                    root,
+                    name: "/".into(),
+                }),
+            })
+    }
```
It feels like maybe this should actually be checking if (pseudocode ahead) `join_path(base_dir, group, "cgroup.procs")` is a file?

bruceg replied:
There are three standard operating modes for cgroups:
1. Legacy v1 mounted at `/sys/fs/cgroup`.
2. Modern v2 mounted at `/sys/fs/cgroup`.
3. Hybrid, with v1 mounted at `/sys/fs/cgroup` and v2 mounted at `/sys/fs/cgroup/unified`.
Cgroups v2 will always have a `cgroup.procs` file in the root directory (based on what I am seeing on all the systems I can access). Cgroups v1 will never have that file. So, the presence of that file is enough to prove the presence of v2 in either the base or `unified` subdirectory.

blt commented on website/cue/reference/components/sources/host_metrics.cue:
```diff
@@ -116,6 +117,72 @@ components: sources: host_metrics: {
 				unit:    "seconds"
 			}
 		}
+		cgroups: {
```
We might want to specifically state what our cgroup support is. Which minimal Linux version do we require? Does this work for v1 or v2 cgroups? Stuff like that.

blt commented on src/sources/host_metrics/cgroups.rs:
```diff
@@ -0,0 +1,271 @@
+            cgroup.load_cpu(),
+            "Failed to load/parse cgroups CPU statistics",
+        ) {
+            result.push(self.counter(
+                "cgroups_cpu_usage",
+                now,
+                cpu.usage_usec as f64 / 1_000_000.0,
+                tags.clone(),
+            ));
+            result.push(self.counter(
+                "cgroups_cpu_user",
+                now,
+                cpu.user_usec as f64 / 1_000_000.0,
+                tags.clone(),
+            ));
+            result.push(self.counter(
+                "cgroups_cpu_system",
+                now,
+                cpu.system_usec as f64 / 1_000_000.0,
+                tags.clone(),
+            ));
```
Out of curiosity why not report the native numbers we receive here?

bruceg replied:
Basically because "seconds" is a base unit (unscaled), and they have to be converted into `f64` anyways.

blt replied:
Ah, the counter itself is in seconds unit?

bruceg replied:
The underlying counter reported by the kernel is in microseconds. I decided to report it to consumers as seconds to match existing precedent (`cpu_seconds_total`).

blt replied:
Gotcha. Absent other limitations my preference as a user would be to get the value without manipulation.

blt commented on src/sources/host_metrics/cgroups.rs:
```diff
@@ -0,0 +1,271 @@
+use chrono::{DateTime, Utc};
+use serde::{Deserialize, Serialize};
+use shared::btreemap;
+use std::fs::{self, File};
+use std::io::{self, Read};
```
By using `std::fs` and `std::io` we're missing async and sync code. The caller will be the async source and its dependencies ought to be async too. For the most part it won't matter except in those cases where the kernel is delayed in handing us back a file handler for cgroup stuff and the source sort mysteriously pauses for a second.

bruceg replied:
Can this delay actually ever happen? That is, is it worth having the constant overhead, admittedly small, of going through the async runtime in case the kernel will stall out one of the steps, when that stall is vanishingly rare in real operation (beyond the normal scheduling micro-delays)? I am open to rewriting it to full async code and we do _indirectly_ pull in `async-std` so it should not be a difficult task at all, it just seems like a waste of code.

blt replied:
> Can this delay actually ever happen? 

Sure. Consider what happens when we're close to the file descriptor limit or are running under a profiler that instruments file opens. In the first case we may see open syscalls appreciably delayed and in the later we will see them slow depending on the intercept. This later is maybe more concerning because it'll mis-instrument vector when running under a profiler. 

With regard to the potential overhead of going through the runtime versus profiling if that's a concern here I'd be happy to reconsider if there are benchmarks to show that overhead's significant. I'm more concerned with consistency of behavior. I have a strong suspicion that the stalling we see in vector in practice -- where capacity drops off a cliff and then picks back up -- is down to the mixing of sync and async code. Very hard to actually debug/diagnose and we can really only catch it when vector's all-up. If we push to make things as async as possible that inconsistency in behavior goes away, is my thinking.

> I am open to rewriting it to full async code and we do indirectly pull in async-std so it should not be a difficult task at all, it just seems like a waste of code.

Why would we use `async-std`? Tokio has filesystem handling code.

bruceg replied:
Got it. I'll do that.

bruceg replied:
Resolved in f2a7eee42

blt commented on src/sources/host_metrics/cgroups.rs:
```diff
@@ -0,0 +1,271 @@
+        join_path(&self.root, filename)
+    }
+
+    fn load_memory_current(&self) -> io::Result<u64> {
+        let mut current = String::new();
+        File::open(self.make_path("memory.current"))?.read_to_string(&mut current)?;
+        current.trim().parse().map_err(map_parse_error)
+    }
+
+    fn load_memory_stat(&self) -> io::Result<MemoryStat> {
+        let mut data = String::new();
+        File::open(self.make_path("memory.stat"))?.read_to_string(&mut data)?;
+        data.parse().map_err(map_parse_error)
+    }
```
To my comment [here](https://github.com/timberio/vector/pull/8806/files#r694121728) these ought to be async based on who their eventual caller is. Eliding the temporary string allocation by parsing from the `Read` directly would also be cool. IIRC this format is intended for such things.

bruceg replied:
Making it all async gets complicated because the cgroups structure is recursive. Recursive async functions are not allowed. So, we would have to drop support for reporting anything beyond the base requested group, or handle the hierarchy by implementing the recursion stack manually. I have implemented reusing the buffer in aadd968ee.

bruceg committed:
Scale microseconds using multiply instead of divide

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

jszwedko reviewed (commented):
Thanks for this @bruceg ! I'd like to test it out, but I left some preliminary comments below.

bruceg commented on website/cue/reference/components/sources/host_metrics.cue:
```diff
@@ -366,6 +433,14 @@ components: sources: host_metrics: {
 		}
 
+		// Host cgroups
+		cgroups_cpu_usage:      _host & _cgroups_cpu & {description:    "The total amount CPU time used by this cgroup and its descendants, in seconds."}
```
Ah, yes, of course.

bruceg replied:
Resolved in 9029afca2

bruceg committed:
Fix metric naming

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

binarylogic commented on src/sources/host_metrics/cgroups.rs:
```diff
@@ -0,0 +1,271 @@
+        let mut result = Vec::new();
+
+        let tags = btreemap! {
+            "cgroup" => cgroup.name.to_string_lossy(),
```
This makes me wonder if adding a `cgroup` tag to the existing metrics is enough to satisfy the requirements? In other words, dropping the `cgroups_*` metrics. Is that possible, or is there any reason you didn't consider this?

bruceg replied:
I am unclear what you are asking here. Is this a question about the naming of the new metrics or of procedure in how the metrics are being generated?

binarylogic replied:
Procedure. Basically, I view cgroup information as a piece of context to the existing host-level metrics. If possible, it should be a facet to slice the existing metrics across, not an entirely new set of metrics. Does that make sense?

bruceg replied:
It does, somewhat. We could potentially replace some of the other metrics gathering procedures with this cgroups module, but it only applies to some of the metrics (ex filesystem space, load average, and network usage are not present), only works on Linux (as opposed to our current cross-platform support), and only when v2 or hybrid cgroups are enabled (there are still some v1-only setups). I am not sure the reported numbers are identical either, though they should be close at least.

binarylogic replied:
Got it. It sounds like it would be somewhat awkward to try and accomplish that, right? Mostly because we rely on a library to collect those metrics.

bruceg replied:
Yes, it would be awkward, at best. More so because we are trying to be as cross-platform as possible.

bruceg committed:
Add note about support limited to v2 or hybrid cgroups.

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

jszwedko commented on website/cue/reference/components/sources/host_metrics.cue:
```diff
@@ -116,6 +117,72 @@ components: sources: host_metrics: {
 		}
+		cgroups: {
+			common:      false
+			description: #"Options for the "cgroups" metrics collector. This collector is only available on Linux systems, as the cgroups (control groups) system is only present on Linux, and only supports either version 2 or hybrid cgroups."#
```
```suggestion
			description: #"""
			Options for the "cgroups" metrics collector.
			
			Note: this collector is only available on Linux systems and only supports either version 2 or hybrid cgroups.
			"""#
```

I think this is slightly easier to read.

bruceg replied:
Applied in 0a55672e7

bruceg committed:
Add commentary on determining the cgroups root directory

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

bruceg commented on website/cue/reference/components/sources/host_metrics.cue:
```diff
@@ -116,6 +117,72 @@ components: sources: host_metrics: {
+				base: {
+					common:      false
+					required:    false
+					description: "The base cgroup name to provide metrics for"
```
It is possible, but I went for the simple case first.

jszwedko replied:
Sure, I guess I'm just wondering how likely it is that users will have multiple cgroups they'll want to track. If we anticipate changing it soon, it'd probably be worth just doing it now to avoid the deprecation dance.

bruceg replied:
I have no good notion if users will want to report multiple cgroups beyond the hierarchy. Doing it will take some work, as there is the potential for one named group to be inside another named group, and we wouldn't want to report the metrics multiple times. On the other hand, recursing the entire groups tree just to find the named groups is wasted effort, so filtering could be tricky.

bruceg committed:
Update top-level config wording.

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

jszwedko commented on website/cue/reference/components/sources/host_metrics.cue:
```diff
@@ -366,6 +437,14 @@ components: sources: host_metrics: {
 
+		// Host cgroups
+		cgroup_cpu_usage_seconds_total:  _host & _cgroup_cpu & {description:    "The total amount CPU time used by this cgroup and its descendants, in seconds."}
+		cgroup_cpu_user_seconds_total:   _host & _cgroup_cpu & {description:    "The amount of CPU time spent in user space, in seconds."}
+		cgroup_cpu_system_seconds_total: _host & _cgroup_cpu & {description:    "The amount of CPU time spent in system tasks, in seconds."}
```
```suggestion
		cgroup_cpu_user_seconds_total:   _host & _cgroup_cpu & {description:    "The total amount of CPU time spent in user space, in seconds."}
		cgroup_cpu_system_seconds_total: _host & _cgroup_cpu & {description:    "The total amount of CPU time spent in system tasks, in seconds."}
```

for consistency

bruceg replied:
Addressed in c1d99b1e9

bruceg committed:
Tweak metric descriptions

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

bruceg committed:
Re-use the parse buffer between I/O calls

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

bruceg commented on website/cue/reference/components/sources/host_metrics.cue:
```diff
@@ -116,6 +117,72 @@ components: sources: host_metrics: {
+					description: "The number of levels of the cgroups hierarchy for which to report metrics. A value of `1` means just the root or named cgroup."
+					type: uint: {
+						unit:    null
+						default: 3
```
I had copied that from `systemd-cgtop`, thinking that it might be similar in other tools. I see now, however, that nothing else has the same limit of three levels, so I am thinking either `0` (no recursion, only the named group) or `u64::MAX` (full recursion) would be more appropriate. Any opinions on which way to go on this?

bruceg committed:
Fix repeated allocation of vectors in cgroups

This should be applied to all the top-level host metrics collectors

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

bruceg committed:
Convert all cgroups metrics functions to async

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

bruceg committed:
Unbreak non-Linux builds

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

jszwedko reviewed (approved):
Just one note about defaulting to full recursion, but otherwise 👍

bruceg committed:
Change default to unlimited cgroup recursion

Is there really no way to indicate the default is unlimited in Cue?

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

bruceg committed:
Change default cgroup levels to fix test failures

The new default should be large enough to encompass any reasonable
cgroup hierarchy, but will limit runaway recursion if that is ever a
concern.

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>
